data-science-onramp/data-cleaning/clean.py

```diff
@@ -0,0 +1,183 @@
+import sys
+import re
+import datetime
+
+from py4j.protocol import Py4JJavaError
+from pyspark.sql import SparkSession
+from pyspark.sql.functions import UserDefinedFunction
+from pyspark.sql.types import StringType, IntegerType, FloatType
+
+
+PROJECT_ID = sys.argv[1]
+BUCKET_NAME = sys.argv[2]
+TABLE = f'{PROJECT_ID}.new_york_citibike_trips.RAW_DATA'
+
+def trip_duration_udf(duration):
+    '''Convert trip duration to seconds.'''
+    if not duration:
+        return None
+    
+    time = re.match('\d*.\d', duration)
+
+    if not time:
+        return None
+
+    time = float(time[0])
+
+    if time < 0:
+        return None
+
+    if 'm' in duration:
+        time *= 60
+    elif 'h' in duration:
+        time *= 60 * 60
+    
+    return int(time)
+
+def station_name_udf(name):
+    '''Replaces '/' with '&'.'''
+    return name.replace('/', '&') if name else None
+
+def user_type_udf(user):
+    '''Converts user type to 'Subscriber' or 'Customer'.'''
+    if not user:
+        return None
+    
+    if user.lower().startswith('sub'):
+        return 'Subscriber'
+    elif user.lower().startswith('cust'):
+        return 'Customer'
+
+def gender_udf(gender):
+    '''Converts gender to 'Male' or 'Female'.'''
+    if not gender:
+        return None
+    
+    if gender.lower().startswith('m'):
+        return 'Male'
+    elif gender.lower().startswith('f'):
+        return 'Female'
+
+def angle_udf(angle):
+    '''Converts DMS notation to angles.'''
+    if not angle:
+        return None
+    
+    dms = re.match('(-?\d*).(-?\d*)\'(-?\d*)"', angle)
+    if dms:
+        return int(dms[1]) + int(dms[2])/60 + int(dms[3])/(60 * 60)
+    
+    try:
+        return float(angle)
+    except ValueError:
```

Comment:
Per our conversation, this probably needs to be expanded to catch other cases that can be converted

If you do use an error, add a short comment explaining why `None` is returned, or add logging.